meta/saimetadatautils.c

```diff
@@ -127,7 +146,7 @@ const sai_attr_metadata_t* sai_metadata_get_attr_metadata_by_attr_id_name(
     {
         ssize_t middle = (first + last) / 2;
 
-        int res = strcmp(attr_id_name, sai_metadata_attr_sorted_by_id_name[middle]->attridname);
+        int res = sai_metadata_attr_id_name_cmp(attr_id_name, sai_metadata_attr_sorted_by_id_name[middle]->attridname, strlen(sai_metadata_attr_sorted_by_id_name[middle]->attridname) + 1);
```

Comment:
why this needs to be updated ?

Reply:
please create separate function for this like sai_metadata_get_attr_metadata_by_attr_id_name_ext, since sai_metadata_get_attr_metadata_by_attr_id_name is performance critical, and you are doing a lot of stuff here for each character in the attribute name and also strlen

Reply:
First of all, one attribute id name may be substring of another attribute id name:
    .attridname                    = "SAI_PORT_ATTR_ADVERTISED_FEC_MODE",
    .attridname                    = "SAI_PORT_ATTR_ADVERTISED_FEC_MODE_EXTENDED",
Secondly, when deserializing an attribute, the input buffer looks like this:
"{\"id\":\"SAI_SWITCH_ATTR_NUMBER_OF_ACTIVE_PORTS\",\"value\":{\"u32\":3}}"
The deserialize function then pass the pointer of the first character of the id name to the get meta by name function. If it is using strcmp, the attribute id will not be found because the buffer is much longer than just attribute id name string. If it is using strncmp, there is a chance that the length of the substring attribute id is used, so that it will match its longer one.
Will create a separate function for this and do refactor to make it more efficient by avoiding the strlen calling.

Reply:
substring dont matter, since strcmp will compare to zero at the end of string so it will always choose correct attribute

Reply:
Yes, usually strcmp works well if input attribute id name is "SAI_SWITCH_ATTR_NUMBER_OF_ACTIVE_PORTS".
However, when deserializing an attribute, the input attribute id name is "SAI_SWITCH_ATTR_NUMBER_OF_ACTIVE_PORTS","value":{"u32":3}}", so that we need a special version of strcmp which takes all the allowed char as '\0'.
Another way is that first extract and copy the attribute id into a separate buffer, then pass it to the get meta by name function. This effects the performance since an additional memory allocation and copy is needed.

Reply:
yes, then i proposed to add _ext version of this api especially for that, since i want existing version to be fast

Reply:
ACK,